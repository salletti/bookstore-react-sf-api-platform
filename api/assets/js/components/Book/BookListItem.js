import React from "react";
import {Link} from "react-router-dom";

//pippo
export default class BookListItem extends React.Component {
  constructor (props) {
    super(props);

    this.controller = new AbortController();
    this.signal = this.controller.signal;

    this._isMounted = false;

    this.state = {
      isHovered: false,
      imagePath: ''
    };

    this.handleHover = this.handleHover.bind(this);
  }

  handleHover () {
    this.setState(prevState => ({
      isHovered: !prevState.isHovered
    }));
  }

  componentDidMount () {
    fetch(this.props.image, {signal: this.signal}).then(response => response.json())
      .then(data => {
        this.setState({imagePath: data.contentUrl});
      })
      .catch(error => {
        if (error.name === 'AbortError') return;
        throw error;
      });
  }

  componentWillUnmount () {
    this.controller.abort();
  }

  render () {
    const isActive = this.state.isHovered ? "active" : "";

    return (
      <li className={`d-flex justify-content-between align-items-center list-group-item ${isActive}`} onMouseEnter={this.handleHover}
          onMouseLeave={this.handleHover}>
        <Link to={`/bookstore/book/${this.props.id}`}>
          {this.props.title} <br/>
          <small>by {this.props.author}</small> <br/>
          <small>{this.props.commentsNumber} avis</small>
        </Link>
        <div>
          <img className="image-item" src={this.state.imagePath} alt={this.state.imagePath}/>
        </div>
      </li>
    );
  }
}
