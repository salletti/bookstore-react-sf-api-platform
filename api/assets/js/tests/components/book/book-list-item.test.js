import React from "react";
import renderer from 'react-test-renderer';

import {BrowserRouter as Router} from "react-router-dom";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

import BookListItem from "../../../components/Book/BookListItem";

let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const bookData = {
  id: 1,
  key: 1,
  title: "title 1",
  author: "author 1",
  commentsNumber: 10,
  image: '/media_objects/1',
};

const fakeImage = {
  id: "/media_objects/1",
  type: "http://schema.org/MediaObject",
  contentUrl: "/media/cover1.jpg",
};

it("show item", async () => {
  //mocking fetch
  global.fetch = jest.fn().mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(fakeImage)
    })
  );

  // Use the asynchronous version of act to apply resolved promises
  await act(async () => {
    render(
      <Router>
        <BookListItem {...bookData} />
      </Router>
      , container);
  });

  expect(container.querySelector("small").textContent).toBe('by ' + bookData.author);
  expect(container.querySelector("small:nth-child(2)").textContent).toBe('by ' + bookData.author);
  expect(container.querySelector("a").getAttribute("href")).toBe(`/bookstore/book/${bookData.id}`);
  expect(container.querySelector("img").getAttribute("src")).toBe("/media/cover1.jpg");

  //destroy mock
  global.fetch.mockClear();
  delete global.fetch;
});

it('item render correctly', () => {
  //mocking fetch
  global.fetch = jest.fn().mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(fakeImage)
    })
  );

  const tree = renderer
    .create(<Router>
      <BookListItem {...bookData} />
    </Router>)
    .toJSON();
  expect(tree).toMatchSnapshot();

  global.fetch.mockClear();
  delete global.fetch;
});
