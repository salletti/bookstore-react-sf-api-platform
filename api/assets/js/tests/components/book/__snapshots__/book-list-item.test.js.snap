// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`item render correctly 1`] = `
<li
  className="d-flex justify-content-between align-items-center list-group-item"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <a
    href="/bookstore/book/1"
    onClick={[Function]}
  >
    title 1
     
    <br />
    <small>
      by 
      author 1
    </small>
     
    <br />
    <small>
      10
       avis
    </small>
  </a>
  <div>
    <img
      alt=""
      className="image-item"
      src=""
    />
  </div>
</li>
`;
